cloudinit/sources/DataSourceCloudStack.py

```diff
@@ -282,7 +289,16 @@ def get_vr_address():
         return latest_address
 
     # Try dhcp lease files next
-    lease_file = dhcp.IscDhclient.get_latest_lease()
+    # get_latest_lease() needs a Distro object to know which directory
+    # stores lease files
+    lease_file = None
+    if distro:
+        lease_file = dhcp.IscDhclient.get_latest_lease(
+            distro.dhclient_lease_directory, distro.dhclient_lease_file_regex
+        )
+    else:
```

Comment:
I'm not sure this should happen at all

Reply:
Hi @igalic, 

Thank you for your comments. I believe all three your comments can be grouped here together in one notion: "explicit is better than implicit, better return an error than continuing with undefined behavior".

I normally would agree with it but there is one more last resort down the code: look for Cloudstack Metadata server at server's default gateway. So even if something is wrong with the DHCP client code, there's still a chance for cloud-init to get its hands on metadata.

I know its hard to discuss the logical flow without knowing whats going on underneath. That's why I explained how Cloudstack networks work in detail in my previous PR: https://github.com/canonical/cloud-init/pull/4433#discussion_r1366978013

Basically the reason behind not raising an exception here is to allow the code to use the last resort, the default gateway address.

Reply:
I don't doubt that you understand the CloudStack code way better than I do, so let's reformulate my change request into two clear and concise questions:

1. under which circumstances can this `distro` object be `None`?
2. why is `distro=None` a desirable state for a `DataSourceCloudStack` instance to be in?

Reply:
1. I don't have an answer to this question. I found that None is passed into DataSourceCloudStack.__init__() during tests so I thought it is better to include a check for it in case None is somehow passed in.
I would assume that distro=None should normally never happen.

2. It is not a desirable state.

I guess if the crash during distro=None is acceptable, I could assume that distro is always provided, and adjust the tests to be initialized with `distros.ubuntu.Distro`  instead of None.

Reply:
Updated the PR: removed the distro=None checks, as suggested.

Reply:
if it's only happening during tests we should just fix the tests